File: 0303.区域和检索-数组不可变/0303-区域和检索-数组不可变.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums
        def x(arr,pre = 0):
            ret = [0]*len(arr)
            if len(arr) == 0:
                return ret
            ret[0] = arr[0]
            for i,num in enumerate(arr[1:], start=1):
                ret[i] = num + ret[i-1]
            return ret
        self.cache = x(self.nums)
            
        

    def sumRange(self, i: int, j: int) -> int:
        return self.cache[j] - self.cache[i-1] if i != 0 else self.cache[j]

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)

File: 0303.区域和检索-数组不可变/0303-区域和检索-数组不可变.rs
struct NumArray {
    nums: Vec<i32>,
    cache: Vec<i32>
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumArray {

    fn new(nums: Vec<i32>) -> Self {
        let nums = nums.clone();
        let mut cache: Vec<i32> = Vec::with_capacity(nums.len()+1);
        cache.push(0);
        for i in 0..nums.len() {
            cache.push(cache[i]+nums[i]);
        }
        
        NumArray {
            nums, cache
        }
    }
    
    fn sum_range(&self, i: i32, j: i32) -> i32 {
        self.cache[(j+1) as usize] - self.cache[i as usize]
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * let obj = NumArray::new(nums);
 * let ret_1: i32 = obj.sum_range(i, j);
 */